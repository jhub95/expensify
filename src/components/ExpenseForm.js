import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';



export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            errorMessage: '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false
        };
    };
    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(()=>({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match( /^\d{1,}(\.\d{0,2})?$/) ){
            this.setState(()=>({amount}));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(()=>({createdAt}));
        }
        
    };
    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused:focused}));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({errorMessage:'please provide description and amount'}));
        } else {
            this.setState(()=>({errorMessage:''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <form onSubmit={this.onSubmit} className="form">
                {this.state.errorMessage && <p className="form__error">{this.state.errorMessage}</p>}
                <input 
                    name="description"
                    id='description'
                    type="text" 
                    className="text-input"
                    placeholder="Description" 
                    value={this.state.description}
                    autoFocus
                    onChange={this.onDescriptionChange}
                />
                <input 
                    name="amount" 
                    id='amount'
                    type="text" 
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=> false }
                />
                <textarea 
                    name="note"  
                    id='note'
                    placeholder="note"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                ></textarea>
                <div>
                    <button type="submit" className="button">submit</button>
                </div>
            </form>
        )
    }
}