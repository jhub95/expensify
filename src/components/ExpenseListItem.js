import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = (props) => 
{
    return (
    <Link to={`/edit/${props.id}`} className="list-item">
        <div>
            <h4 className="list-item__title">{props.description}</h4>
            <span className="list-item__subtitle">{moment(props.createdAt).format('MMMM Do, YYYY')}</span>
        </div>
       <h4 className="list-item__data">{numeral(props.amount/100).format('$0,0.00')}</h4>
        
    
    </Link>
    
    
    
)};

export default ExpenseListItem;