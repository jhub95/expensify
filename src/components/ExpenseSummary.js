import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from './../selectors/expenses';
import getExpensesTotal from './../selectors/expenses-total';
import numeral from 'numeral';

export const ExpenseSummary = ({expenses})=>{
    const length = expenses.length;
    const total = getExpensesTotal(expenses);
    if (length > 0){
        return (
        <p>Viewing {length} {length > 1 || length === 0 ? 'expenses' : 'expense'} totalling {numeral(total/100).format('$0,0.00')}</p>   
        )
    } else {
        return null;
    }
};

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);