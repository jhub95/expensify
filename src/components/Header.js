import React from 'react';
import {Link,NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <ul>
            <NavLink to="/" activeClassName="is-active" exact={true}><li>Home</li></NavLink>
            <NavLink to="/create" activeClassName="is-active"><li>Create New Expense</li></NavLink>
            <NavLink to="/help" activeClassName="is-active"><li>Help Page</li></NavLink>
        </ul>
    </header>
);

export default Header;