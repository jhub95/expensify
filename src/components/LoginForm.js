import React from "react";
import { connect } from 'react-redux';
import { startLogin } from './../actions/auth';

// import {firebase, googleAuthProvider} from './../firebase/firebase';

export const LoginForm = ({startLogin})=>{
    return (
        <div>
        <h1>Please login to see your expenses.</h1>
            <form>
                <input 
                    type='text' 
                    id='login_name'
                    name='login_name' 
                    placeholder="username"
                />
                <input 
                    type="password" 
                    id='password' 
                    name='password'
                    placeholder="password"
                />
                
            </form>
            <button id='loginButton' onClick={startLogin}>Login</button>
        </div>
    );

};

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginForm);