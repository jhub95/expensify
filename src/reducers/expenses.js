const expensesReducerDefaultState=[];

export const expensesReducer = (
    state=expensesReducerDefaultState, 
    {type,expense,id,updates}=action
) =>{
    switch (type) {
     case 'ADD_EXPENSE':
        return [
            ...state,
            expense
        ];
     case 'REMOVE_EXPENSE':
        return state.filter((tempExpense)=>tempExpense.id !== id);
     case 'EDIT_EXPENSE':
        return state.map((expense)=>{
           if (expense.id === id){
            return {
                ...expense,
                ...updates
            };
           } else {
               return expense;
           }
        });
     default:
        return state;
    }
};