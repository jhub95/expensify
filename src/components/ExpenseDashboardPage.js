import React from 'react';
import { connect } from 'react-redux';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';
import selectExpenses from './../selectors/expenses';

export const ExpenseDashboardPage = (props) => {
    console.log(props);
    return(
        <div>
        <ExpenseSummary />
        {props.expenses.length > 0 ? <ExpenseListFilters /> : ''}
        <ExpenseList />
        </div>
    );
}

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseDashboardPage);