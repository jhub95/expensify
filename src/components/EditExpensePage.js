import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, startRemoveExpense} from './../actions/expenses';

export class EditExpensePage extends React.Component{
    startEditExpense = (expense) => {
        console.log("expenseEditPage: expenses-",expense)
        this.props.startEditExpense(this.props.expense.id,expense);
        this.props.history.push('/');
    };
    removeExpense = (e) => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };
    render(){
        return (
        <div>
        <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.startEditExpense}
        />
        <button onClick={this.removeExpense}
        >remove</button>
        </div>
        )
    };
};

const mapStateToProps = (state,props) => ({
    expense:state.expenses.find((expense)=> expense.id === props.match.params.id) 
});

const mapDispatchToProps = (dispatch,props)=>({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);