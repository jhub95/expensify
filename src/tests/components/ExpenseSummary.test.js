import React from 'react';
import expenses from './../fixtures/expenses';
import {ExpenseSummary} from './../../components/ExpenseSummary';
import {shallow} from 'enzyme';

describe('from components/ExpenseSummary',()=>{

    test('should return null if no expenses',()=>{
        const wrapper = shallow(<ExpenseSummary expenses={[]} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should return declaration of expenses count and sum',()=>{
        const wrapper = shallow(<ExpenseSummary expenses={expenses} />);
        expect(wrapper).toMatchSnapshot();
    });

});