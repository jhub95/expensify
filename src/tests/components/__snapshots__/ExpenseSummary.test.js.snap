// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`from components/ExpenseSummary should return declaration of expenses count and sum 1`] = `
<p>
  Viewing 
  3
   
  expenses
   totalling 
  $1,141.95
</p>
`;

exports[`from components/ExpenseSummary should return null if no expenses 1`] = `""`;
